### ISSUE TITLE
Device name (Fiji <-> gfx803) in ROCm 1.6

### ISSUE THREAD

**Issue by newling** [CNugteren/CLBlast#186]
Hello,

The device name for AMD Fiji Nano GPU on ROCm is gfx803. CLBlast therefore returns "default" 
kernels on ROCm for this device.

For an idea of what devices are called in ROCm, here's how MIOpen maps device names (not 1-to-1)

https://github.com/ROCmSoftwarePlatform/MIOpen/blob/9fb1826dd3694d6a4c88ab96e760c4d25b226559/src/include/miopen/device_name.hpp#L44

Thanks.

**Comment by newling** [comment]
As a side comment, on ROCm 1.6 performance seems better with database entry "default" than "Fiji". 
For m=n=k=1024:

"Fiji"       ~ 1.5 teraflops
"default" ~ 2.0 teraflops

I guess this is the new compiler behaving differently to the Catalyst compiler. I will run the tuner and report back.

**Comment by newling** [comment]
With tuning I am getting improved results (~2.5 teraflops) 
[ -------> ] 0.87 ms or 2481.1 GFLOPS
but this might only be after partial tuning (see issue #187)
This time appears to be better than on Catalyst, from 
https://github.com/CNugteren/CLBlast-database/database.json
I see 2.09 ms (~1 teraflop)

**Comment by CNugteren** [comment]
OK, thanks for that link! I'm actually planning to introduce a device name mapping system in CLBlast as well, since this issue occurs more often. Plus, that way we can unify tuning results for all kinds of similar devices (e.g. for Intel CPUs or NVIDIA GPUs of the same architecture).

I'll include this new ROCm mapping there as well.

**Comment by newling** [comment]
Sounds good.

Running host-side timer on SGEMM 1024 x 1024 x 1024 
(Fiji Nano on ROCm 1.6)

- Tuned for device: 
total host time : 37.3572 [ms] total runs : 47 mean gflops : 2701.8

- Using AMD default from database: 
total host time : 52.0262 [ms] total runs : 47  mean gflops : 1940.02

- Using AMD Fiji from database (I guess tuned with Catayst):
total host time : 73.3061 [ms] total runs : 47  mean gflops : 1376.85

This could be a case from Issue #169 where tuning for a device is not specific enough.

**Comment by CNugteren** [comment]
About the ROCm mapping, it is even more high-level than before! So now there is no easy way to distinguish a Polaris GPU from a Fiji GPU, because they are all called `gfx803`? See also here:
http://llvm.org/docs/AMDGPUUsage.html

Seems `CL_DEVICE_NAME` is becoming the architecture name at AMD. Do you know whether a tool like `clinfo` prints more detailed info (e.g. `Radeon R9 Nano`)?

**Comment by newling** [comment]
That's interesting. `clinfo` does print more detailed info 

```
/opt/rocm/opencl/bin/x86_64/clinfo
```

gives

```
Number of platforms:				 1
  Platform Profile:				 FULL_PROFILE
  Platform Version:				 OpenCL 2.0 AMD-APP (2450.0)
  Platform Name:				 AMD Accelerated Parallel Processing
  Platform Vendor:				 Advanced Micro Devices, Inc.
  Platform Extensions:				 cl_khr_icd cl_amd_event_callback cl_amd_offline_devices 


  Platform Name:				 AMD Accelerated Parallel Processing
Number of devices:				 1
  Device Type:					 CL_DEVICE_TYPE_GPU
  Vendor ID:					 1002h
  Board name:					 Fiji [Radeon R9 FURY / NANO Series]
  Device Topology:				 PCI[ B#2, D#0, F#0 ]
  Max compute units:				 64
  Max work items dimensions:			 3
    Max work items[0]:				 256
    Max work items[1]:				 256
    Max work items[2]:				 256
  Max work group size:				 256
  Preferred vector width char:			 4
  Preferred vector width short:			 2
  Preferred vector width int:			 1
  Preferred vector width long:			 1
  Preferred vector width float:			 1
  Preferred vector width double:		 1
  Native vector width char:			 4
  Native vector width short:			 2
  Native vector width int:			 1
  Native vector width long:			 1
  Native vector width float:			 1
  Native vector width double:			 1
  Max clock frequency:				 1000Mhz
  Address bits:					 64
  Max memory allocation:			 3221225472
  Image support:				 Yes
  Max number of images read arguments:		 128
  Max number of images write arguments:		 8
  Max image 2D width:				 16384
  Max image 2D height:				 16384
  Max image 3D width:				 2048
  Max image 3D height:				 2048
  Max image 3D depth:				 2048
  Max samplers within kernel:			 29440
  Max size of kernel argument:			 1024
  Alignment (bits) of base address:		 1024
  Minimum alignment (bytes) for any datatype:	 128
  Single precision floating point capability
    Denorms:					 No
    Quiet NaNs:					 Yes
    Round to nearest even:			 Yes
    Round to zero:				 Yes
    Round to +ve and infinity:			 Yes
    IEEE754-2008 fused multiply-add:		 Yes
  Cache type:					 Read/Write
  Cache line size:				 64
  Cache size:					 16384
  Global memory size:				 4294967296
  Constant buffer size:				 3221225472
  Max number of constant args:			 8
  Local memory type:				 Scratchpad
  Local memory size:				 65536
  Max pipe arguments:				 0
  Max pipe active reservations:			 0
  Max pipe packet size:				 0
  Max global variable size:			 3221225472
  Max global variable preferred total size:	 4294967296
  Max read/write image args:			 64
  Max on device events:				 0
  Queue on device max size:			 0
  Max on device queues:				 0
  Queue on device preferred size:		 0
  SVM capabilities:				 
    Coarse grain buffer:			 Yes
    Fine grain buffer:				 Yes
    Fine grain system:				 No
    Atomics:					 No
  Preferred platform atomic alignment:		 0
  Preferred global atomic alignment:		 0
  Preferred local atomic alignment:		 0
  Kernel Preferred work group size multiple:	 64
  Error correction support:			 0
  Unified memory for Host and Device:		 0
  Profiling timer resolution:			 1
  Device endianess:				 Little
  Available:					 Yes
  Compiler available:				 Yes
  Execution capabilities:				 
    Execute OpenCL kernels:			 Yes
    Execute native function:			 No
  Queue on Host properties:				 
    Out-of-Order:				 No
    Profiling :					 Yes
  Queue on Device properties:				 
    Out-of-Order:				 No
    Profiling :					 No
  Platform ID:					 0x7f50e2950598
  Name:						 gfx803
  Vendor:					 Advanced Micro Devices, Inc.
  Device OpenCL C version:			 OpenCL C 2.0 
  Driver version:				 1.1 (HSA,LC)
  Profile:					 FULL_PROFILE
  Version:					 OpenCL 1.2 
  Extensions:					 cl_khr_fp64 cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_int64_base_atomics cl_khr_int64_extended_atomics cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_gl_sharing cl_amd_media_ops cl_amd_media_ops2 cl_khr_subgroups cl_khr_depth_images cl_amd_liquid_flash cl_amd_copy_buffer_p2p 
```

I'm not sure how to get
```
Board name:					 Fiji [Radeon R9 FURY / NANO Series]
```
through the OpenCL API though.

**Comment by tmagomedov** [comment]
Maybe (device arch/name, number of compute units, frequency) triples can do the job if there is not other way to distinguish devices.

**Comment by CNugteren** [comment]
Thanks for the info! Yes, there is indeed a `CL_DEVICE_BOARD_NAME_AMD` (part of an [AMD extension](https://www.khronos.org/registry/OpenCL/extensions/amd/cl_amd_device_attribute_query.txt)).

I propose to introduce the following hierarchy (see also issue #169):

1.  **Top-level**: Vendor name (`CL_DEVICE_VENDOR`) plus device type (`CL_DEVICE_TYPE`). Examples are "AMD GPU" and "Intel CPU".
2. **Architecture**: The device family. Examples are "Fiji", "Maxwell" and "Skylake".
3. **Device**: The device model. Examples are "R9 M370X" and "GeForce GTX 1080 Ti".

The architecture and device level will be determined differently per vendor:

* For **AMD**, the `CL_DEVICE_NAME` gives the architecture name and `CL_DEVICE_BOARD_NAME_AMD` gives the device name. Perhaps in combination with the number of compute units if it appears to be insufficient.
* For **NVIDIA**, a combination of `CL_DEVICE_COMPUTE_CAPABILITY_MAJOR_NV` and `CL_DEVICE_COMPUTE_CAPABILITY_MINOR_NV` ([NVIDIA extension](https://www.khronos.org/registry/OpenCL/extensions/nv/cl_nv_device_attribute_query.txt)) gives the architecture 'name' and `CL_DEVICE_NAME` gives the device name.
* For others (unless specified) the architecture level is omitted and the device name is given by `CL_DEVICE_NAME`.

On top of all this, there will be a translation table at each level in case of inconsistent driver-based naming (like the AMD APP versus ROCm naming mentioned in this issue).

One could image a fourth lowest level which includes the driver used (or even the version) to really be able to fine tune based on the compiler. Not sure if this is highest priority now given the already complex task at hand.

Any feedback?

**Comment by newling** [comment]
Both proposals look reasonable to me. I suppose the hierarchical one would make choosing fallback database entries cleaner. 

I don't know whether the 40% improvement I get (see above) over the database entry by doing a fresh tune on ROCm (on a Fiji nano) is because of radically different compilers or something else, but maybe the fourth level is important. Maybe a variable depth hierarchy of sorts would be good.

**Comment by CNugteren** [comment]
I have implemented a first version of the above suggestions (3 level hierarchy + translation table). This should not only fix the `gfx803` name, but also use the name of your board. That makes it possible to have different tuning results for different boards with the same architecture (e.g. Fiji). Right now the tuning database looks like this (small sample):
```
    { // AMD GPUs
      kDeviceTypeGPU, "AMD", {
        { "Fiji", {
          { "AMD Radeon R9 Fury X",                            { 16, 16, 1, 2 } },
          { "AMD Radeon R9 M370X Compute Engine",              { 32, 8, 4, 1 } },
          { "default",                                         { 16, 16, 1, 2 } },
        } },
        { "Hawaii", {
(etc.)
```

When you have time, could you checkout the latest `master` of CLTune and the `database_improvements` branch of CLBlast? You can then re-rerun the tuner or run a CLBlast test program in verbose mode (`-DVERBOSE=ON`) to see how it selects tuning parameters.

There are still some issues on my Linux NVIDIA test device, but perhaps you can already test on AMD hardware? Thanks!

**Comment by newling** [comment]
Hi, I've finally had a chance to try these changes  ... on freshly pulled `master`  of both CLBlast and CLTune.

```
./clblast_tuner_copy_pad
```

produces a json file which looks like this, 

```
{
  "kernel_family": "pad",
  "precision": "32",
  "clblast_device_type": "GPU",
  "clblast_device_vendor": "AMD",
  "clblast_device_architecture": "",
  "clblast_device_name": "gfx803",
  "arg_m": "1024",
  "arg_n": "1024",
  "arg_alpha": "2.00",
  "device": "gfx803",
  "platform_version": "OpenCL 2.0 AMD-APP (2450.0)",
  "device_vendor": "Advanced Micro Devices, Inc.",
  "device_type": "GPU",
  "device_core_clock": "1000",
  "device_compute_units": "64",
  "device_extra_info": "",
  "results": [
    {
      "kernel": "CopyPadMatrix",
      "time": 0.131,
      "parameters": {"PAD_DIMX": 8,"PAD_DIMY": 8,"PAD_WPTX": 1,"PAD_WPTY": 1,"PRECISION": 32}
    },
    {
      "kernel": "CopyPadMatrix",
(etc.)
```

I then try 

```
python ../scripts/database/database.py . ..
```

which gives an error `KeyError: 'parameter_names'` 

does the above script only work if all tuners have been run, as in the README file?

**Comment by CNugteren** [comment]
Hmmm, that's not good. The information you got:
```
  "clblast_device_type": "GPU",
  "clblast_device_vendor": "AMD",
  "clblast_device_architecture": "",
  "clblast_device_name": "gfx803",
```
Is not what I expected. Ideally the device name `gfx803` would become the architecture name and then that would be translated to `Fiji` with some look-up table. The device name itself should be filled with the AMD board name.

My guess why this doesn't work (I don't have access to AMD hardware myself right now) is that CLBlast triggers this behaviour only for devices with the `cl_amd_device_attribute_query` extension, because that means that there is `CL_DEVICE_BOARD_NAME_AMD` to query. From the clinfo result you posted earlier it seems you don't have `cl_amd_device_attribute_query` defined. Why would that be?

About the database issue... a lot has changed in the meantime, so perhaps you should remove your local copy of `scripts/database/database.json` and then the script will automatically re-download a new version. If that still doesn't work, send me the tuner output and I'll investigate.

**Comment by newling** [comment]
I don't know anything about `cl_amd_device_attribute_query`, maybe an issue could be raised at https://github.com/RadeonOpenCompute/ROCm

As for the database issue, I deleted `scripts/database/database.json` file and it worked. Thanks.

**Comment by CNugteren** [comment]
I've reported the issue on the ROCm page (see reference). However, I realized just now that your `clinfo` output mentions `OpenCL 2.0 AMD-APP (2450.0)`. Are you sure this is ROCm and not the old AMD APP? Or is this how ROCm identifies itself?

**Comment by newling** [comment]
Thanks. It's definitely the clinfo installed with ROCm `/opt/rocm/opencl/bin/x86_64/clinfo` :)

**Comment by CNugteren** [comment]
Yes, I understand it's ROCm's clinfo, but what I meant was is it the ROCm platform that's reported there? These clinfo programs just scan your system for OpenCL platforms and will report anything from any vendor. But a quick online search indeed reveals that `AMD-APP (2450.0)` seems to be ROCm.

**Comment by CNugteren** [comment]
I've asked again today on the ROCm GitHub (https://github.com/RadeonOpenCompute/ROCm/issues/215) and it should work now in version 1.7. Could you try again and see what your `clinfo` gives you? And then afterwards also see what CLBlast's `clblast_test_diagnostics` prints?

**Comment by newling** [comment]
I don't have access to a computer with ROCm installed anymore, sorry.

**Comment by CNugteren** [comment]
OK, no worries. I'll close this issue for now, because it might well be fixed with ROCm 1.7. If you or anyone else encounters this issue again, feel free to re-open.
